File: 19. Remove Nth Node From End of List.py
'''
Given the head of a linked list, remove the nth node from the end of the list and return its head.
'''
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        stack = []
        current = head
        while current:
            stack.append(current)
            current = current.next
        if n == 1: # remove tail
            if len(stack) > 1:
                stack[-n-1].next = None
            else:
                return None
        elif n == len(stack): # remove head
            if len(stack) > 1:
                return stack[-n+1]
            else:
                return None
        else:
            stack[-n-1].next = stack[-n+1]
        return head

File: 19. Remove Nth Node From End of List.cpp
/*
Given the head of a linked list, remove the nth node from the end of the list and return its head.
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int length = 0;
        ListNode* temp = head;
        while(temp){
            length++;
            temp = temp->next;
        }
        std::cout<<length<<std::endl;

        // remove head base case as n >= 1 and ListNode_length >= 1
        if(length == n){
            return head->next;
        }

        temp = head;
        ListNode* prev = NULL;
        // traverse temp to the node to remove, and prev to the previous node
        for(int i = 0; i < length-n; i++){
            prev = temp;
            temp = temp->next;
        }
        std::cout << prev->val << std::endl;
        // remove temp from the list
        prev->next = temp->next;

        return head;
    }
};